lib/ansible/modules/cloud/docker/docker_swarm_service.py

```diff
@@ -1759,7 +1764,7 @@ def compare(self, os):
         force_update = False
         if self.endpoint_mode is not None and self.endpoint_mode != os.endpoint_mode:
             differences.add('endpoint_mode', parameter=self.endpoint_mode, active=os.endpoint_mode)
-        if self.env is not None and self.env != (os.env or []):
+        if has_list_changed(self.env, os.env):
```

Comment:
The current behavior is:
 * if the user did not specify `env` (i.e. `self.env is None`), don't compare.
 * if the user did specify `env`, compare user's `env` to the service's `env`.
 * if the service's `env` is empty, treat it as `[]`: this simply means that no env variables have been set for the service.

So this does not ignore empty lists, but treats them correctly. (What it doesn't do is consider order - you're totally correct on that one.)

Reply:
A better version would be:
```suggestion
        if has_list_changed(self.env, os.env):
        if self.env is not None and has_list_changed(self.env, os.env or []):
```

Reply:
(Same for the comparisons below.)

Reply:
> The current behavior is:
> 
>     * if the user did not specify `env` (i.e. `self.env is None`), don't compare.
Already implemented (line 1251,1252)
> 
>     * if the user did specify `env`, compare user's `env` to the service's `env`.
Done
> 
>     * if the service's `env` is empty, treat it as `[]`: this simply means that no env variables have been set for the service.
Already implemented (line 1253)
> 
> 
> So this does not ignore empty lists, but treats them correctly. (What it doesn't do is consider order - you're totally correct on that one.)


I believe all the list comparisons seemed to want the following:

  * If the new list is empty, skip comparing
  * If the old list is None, treat it as the empty list
  * Compare the lists. Most should be comparisons between sorted lists.

Can you point out any list comparisons where the above assumptions break? If they do I can adjust them.

Reply:
If the new list is `[]`, you should compare. The user explicitly requests an empty list in that case. If the new list is `None`, don't compare, the user is saying "I don't care".

Reply:
(But that's already what you implemented. So it should be fine.)